File: sem4/mpp/project-cs/networking/Worker/AbstractWorker.cs
using System.Net.Sockets;

namespace networking.Worker;

public abstract class AbstractWorker {
    protected TcpClient connection;

    protected NetworkStream stream;
    protected StreamReader streamReader;
    protected StreamWriter streamWriter;
    protected volatile bool connected;
    protected static Response okResponse;

    public AbstractWorker(TcpClient connection) {
        this.connection = connection;
        okResponse = new Response() {
            type = ResponseType.OK
        };
        try {
            stream = connection.GetStream();
            streamReader = new StreamReader(stream);
            streamWriter = new StreamWriter(stream);
            connected = true;
        }
        catch (Exception e) {
            Console.WriteLine(e.StackTrace);
        }
    }

    public void run() {
        while (connected) {
            try {
                var request = ReadRequest();
                var response = HandleRequest(request);
                if (response != null) {
                    SendResponse(response);
                }
            }
            catch (Exception e) {
                if (!connection.Connected)
                    connected = false;
                Console.WriteLine(e.StackTrace);
            }

            try {
                Thread.Sleep(50);
            }
            catch (Exception e) {
                Console.WriteLine(e.StackTrace);
            }
        }

        try {
            streamReader.Close();
            streamWriter.Close();
            stream.Close();
            connection.Close();
        }
        catch (Exception e) {
            Console.WriteLine("Error " + e);
        }
    }

    protected abstract Request ReadRequest();
    protected abstract Response HandleRequest(Request request);
    protected abstract void SendResponse(Response response);
}

File: sem4/mpp/project-java/networking/src/main/java/networking/worker/AbstractWorker.java
package networking.worker;

import networking.Request;
import networking.Response;
import networking.ResponseType;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.Socket;

public abstract class AbstractWorker implements Runnable {
	protected final Socket connection;

	protected BufferedReader input;
	protected PrintWriter output;
	protected volatile boolean connected;

	protected static Response okResponse;

	public AbstractWorker(Socket connection) {
		this.connection = connection;
		okResponse = new Response();
		okResponse.setType(ResponseType.OK);
		try {
			output = new PrintWriter(connection.getOutputStream());
			input = new BufferedReader(new InputStreamReader(connection.getInputStream()));
			connected = true;
		} catch (IOException e) {
			e.printStackTrace();
		}
	}

	@Override
	public void run() {
		while (connected) {
			try {
				Request request = readRequest();
				Response response = handleRequest(request);
				if (response != null) {
					sendResponse(response);
				}
			} catch (Exception e) {
				if (connection.isClosed())
					connected = false;
				e.printStackTrace();
			}
			try {
				Thread.sleep(50);
			} catch (InterruptedException e) {
				e.printStackTrace();
			}
		}
		try {
			input.close();
			output.close();
			connection.close();
		} catch (IOException e) {
			System.out.println("Error " + e);
		}
	}


	protected abstract Request readRequest() throws IOException;

	protected abstract Response handleRequest(Request request);

	protected abstract void sendResponse(Response response);
}
